Simplify entrypoint script a little bit

d3adb5 opened the pull request:
Refactor the entrypoint script to make it more readable for programmers.

Since variables should be documented in the `README.md`, comments were removed
where I felt they were superfluous, and the ones that felt necessary to identify
sections of the script were shortened, if possible.

As there was an instance of Bash's extended builtin test command (`[[]]`), all
tests that were previously using the POSIX command were switched in favor of it.
The shebang mentions Bash, after all.

Another style choice was inlining a few if statements where possible, and even
using `return` instead of `echo` in `array_contains`. The functions were
shortened, one even "flattened".

Bash's `readonly` and `declare` were introduced.

A couple of items from the ShellCheck output were addressed in 7f3c2c1. The
relevant items are `SC2129` and `SC2145`.

Instead of an overarching `if` statement in the `while` loop that processes
config variables, `grep` was used in its input in 6b48b36.

Finally, trailing whitespace was removed, though this is frowned upon by many.
Thought I would, since there were only three instances of it.

d3adb5 committed:
refactor: eliminate if by grep'ing for MUMBLE_CONFIG_

Use grep to fetch only environment variables that start with
`MUMBLE_CONFIG_`, rather than manually compare in an if statement,
getting rid of one level of indentation.

d3adb5 committed:
refactor: use bash's extended test ([[]])

Replace the standard POSIX test command ([]) with Bash's builtin ([[]])
for consistency.

There was a single use of the double bracket syntax, but since the
script references Bash in its shebang, we might as well use it instead
of the POSIX one.

d3adb5 committed:
refactor: remove trailing whitespace

Remove trailing whitespace while there isn't much. This is discouraged,
since it can mess with git-blame information, but it irks me there is
trailing whitespace.

This commit might be dropped if requested during review. :(

d3adb5 committed:
refactor: inline a couple of if statements

Inline a couple of if statements for a shorter, even if not by much,
script.

d3adb5 committed:
refactor: shorten functions in entrypoint script

Shorten `array_contains` and `set_config` in a way that, in my opinion,
is still readable.

d3adb5 committed:
refactor: remove comment clutter

Remove most comments and shorten some others. Focus should be on making
code clean and immediately readable instead of documented directly in
the script file.

d3adb5 committed:
refactor: declare most relevant variables at the start

Declare the most relevant variables at the start of the script, so the
reader can be made aware of them more quickly.

d3adb5 committed:
refactor: eliminate shellcheck warnings & errors

Redirect multiple commands to the configuration file instead of making
multiple redirects, potentially limiting the number of times the file
needs to be opened for appending. (SC2129)

Use "${arr[*]}" not to mix array and string. (SC2145)

Krzmbrzl commented:
I originally avoided `grep` as it returns a non-zero exit status if it doesn't match anything, which in combination with the `-e` flag would abort the script. To void that some special workarounds are usually necessary. I don't seem to find those in your version. Did you test that a non-zero exit code of `grep` never crashes the script?

Krzmbrzl commented on entrypoint.sh:
```diff
@@ -1,136 +1,104 @@
+			exit 1
 		fi
-	done < <( printenv --null ) # Feeding it in like this, prevents the creation of a subshell for the while-loop
+	done < <( printenv --null | grep -az MUMBLE_CONFIG_ ) # Feeding it in like this, prevents the creation of a subshell for the while-loop
```
What does this do, if there is no `MUMBLE_CONFIG_*` variable defined? Will `grep` cause the script to error due to non-zero exit code?

d3adb5 replied:
While `grep` will have a non-zero exit code, Bash's `set -e` won't break the script because of it. It's the exit code for the `while loop` that really matters, as far as I can tell, or at the very least the exit code for the command that generates the file descriptor seems not to matter for  `set -e`.

The script doesn't error out when a container is launched without any `MUMBLE_CONFIG_*` variables. A quick test shows Bash handles it well too:

```bash
#!/bin/bash
set -e

cat < <(printenv --null | grep -az MUMBLE_CONFIG_)
cat < <(false)
echo test
```

The script above outputs `test`.

Krzmbrzl commented on entrypoint.sh:
```diff
@@ -140,7 +108,7 @@ echo "Running Mumble server as uid=$(id -u) gid=$(id -g)"
 echo "  $( stat ${DATA_DIR} --printf='%A, owner: \"%U\" (UID: %u), group: \"%G\" (GID: %g)' )"
 
-echo "Command run to start the service : ${server_invocation[@]}"
+echo "Command run to start the service : ${server_invocation[*]}"
```
What's the difference between `[@]` and `[*]` in this case?

d3adb5 replied:
For the `echo` command itself it achieves the same behavior, but a linter like ShellCheck will claim this is an error unless we explicitly ignore SC2145. The reason for it is `[@]` and `[*]` behave differently when in the middle of a string:

```bash
declare -a someArray=(1 2 3 4)

# This is how [@] is expanded when in the middle of a string (these command are the same)
echo "my array: ${someArray[@]} yay"
echo "my array: 1" "2" "3" "4 yay"

# This is how [*] is expanded when in the middle of a string (these commands are the same)
echo "my array: ${someArray[*]} yay"
echo "my array: 1 2 3 4 yay"
```

Because `[@]` generates more arguments while `[*]` doesn't, using the former in the middle of a string could have unwanted consequences.

More info here: [SC2145](https://github.com/koalaman/shellcheck/wiki/SC2145)

Krzmbrzl reviewed (commented):
The changes generally LGTM.

However, I am not particularly fond of removing comments. Yes, there exists the documentation in the README, but imo that should not replace in-source documentation in form of comments.

d3adb5 commented:
> The changes generally LGTM.
> 
> However, I am not particularly fond of removing comments. Yes, there exists the documentation in the README, but imo that should not replace in-source documentation in form of comments.

I left a few since I thought they were useful, but before the entrypoint seemed a little overcommented to me. If you still feel they should be present, I can push a commit restoring the previous ones, or drop [ec402cb](https://github.com/mumble-voip/mumble-docker/pull/6/commits/ec402cb89b57c936ee69e5340c12d3a1f3b20029).

azlux commented:
Hi,
I like your commits. It's change a little bit our goal with your changes.
We want, at first, a fully compatible POSIX script, but we face many difficulties with regex (workaround are horrible to read/code). So because we will not do it anymore, your PR is relevant, especially about `readonly`, `declare` , `[[`.

The only change I don't like is removing the comments. As simple dev on my free time, without comments, I need to check a lot the docs to read the script.
Maybe making them less verbose (and thinner), I'm agree with that, but don't remove it  for now.

d3adb5 committed:
refactor: populate existing_config_options with readarray

Populate the array that contains existing configuration options obtained
from the bare-bones configuration file using Bash's readarray builtin.

d3adb5 committed:
docs: restore a few comments removed in ec402cb

Restore a few of the comments removed in ec402cb.

d3adb5 committed:
refactor: use associative array for existing options

Use Bash's associative array for the existing configuration options and
for the translation of something like "DBUSER" to "dbUser", replacing
the for loop that existed before.

d3adb5 commented:
I restored most comments and renamed a variable for good measure in [0f4fd08](https://github.com/mumble-voip/mumble-docker/pull/6/commits/0f4fd08fe47ad6a2923de247b6c670ae7c869ebc).

Since you said Bash is now the target, I went ahead and included `readarray` and even an associative array. I would've liked to keep the RegEx matching to just Bash, but thought it was trading readability for unnecessary performance. Instead, I'm using `sed -E` to retrieve exactly what we want from the bare-bones config file.

Sorry for adding even more code to be reviewed. Hope this isn't too tiresome for a first PR. :sweat_smile:

Krzmbrzl commented:
> Sorry for adding even more code to be reviewed. Hope this isn't too tiresome for a first PR. sweat_smile

No worries - I'll have another deep look at your changes sometime soon-ish :)

Krzmbrzl commented on entrypoint.sh:
```diff
@@ -1,136 +1,102 @@
-			fi
+		# Uppercase and remove underscores so MUMBLE_CONFIG_A_B == MUMBLE_CONFIG_AB
+		uppercase_variable="${var/MUMBLE_CONFIG_/}"
+		config_option="${config_from_uppercase[${uppercase_variable//_/}]}"
```
If I didn't miss anything, I think this right here introduces a possible bug. In case we ever introduced a config-option that does indeed contain an underscore in it, `config_from_uppercase` will also use the upper-cased config's name as a key. _Without_ having removed the underscore. Thus, by unconditionally removing the underscore from `uppercase_variable` _before_ the lookup, we would actually fail to find that config value.

Now, afaik we don't have any configs that use underscores yet, so I think the script should be working fine for all currently existing options. However, you never know what might be added in the future :)

That's why in the original script I first checked whether the upper-cased config name is a known option and only then check if the version with removed underscores is a known option.
That still doesn't cover the case where the original config option contains an underscore and the user added an additional one for readability, but at least when using the exact config name it'd work.

Another alternative would be to apply the same underscore-remove logic to the keys used in `config_from_uppercase` which has the underlying assumption that no two options differ only by casing and underscores. I guess that is a fine assumption though.

d3adb5 replied:
Went with the latter assumption for simplicity in [3f908fd](https://github.com/mumble-voip/mumble-docker/pull/6/commits/3f908fd921883ef8c44ba61168864b501ff99f46). I hadn't realized it and hoped maybe INI files don't allow for keys with underscores, ridiculous as that sounds.

Thanks! :)

Krzmbrzl reviewed (commented):
Otherwise LGTM

d3adb5 committed:
fix: remove underscores from cfg options in assoc array

Remove any possible underscores from the config options when creating
keys for the associative array used to look them up after capturing
environment variables.

This is a possible regression from fc0363a.

Krzmbrzl commented:
Maybe we can increase the readability of the code by introducing a separate function `normalize_config_name` that does the upper-casing and underscore-removing 
And then use that in both places. That would facilitate potential future refactorings

d3adb5 commented:
I went one step further and removed `MUMBLE_CONFIG_` from the beginning as well in the same function. Before I commit, what do you think?

```bash
normalize_name() {
	local uppercase="${1^^}"
	local no_var_prefix="${uppercase#MUMBLE_CONFIG_}"
	echo "${no_var_prefix//_/}"
}
```

The one problem this might cause is config options that contain some case of `mumble_config_` in the beginning of their names. I thought, just like a potential collision of config options differing only in case and/or underscores, it was an easy assumption to make, but I want to know what you think.

With the above, we can just use `${option_for[$(normalize_name "$config")]}` and `${option_for[$(normalize_name "$var")]}`.

Krzmbrzl commented:
I see how this would turn in handy and I certainly think that the assumption that no actual config name will start with mumble_config_ is more than justified.
The only thing that itches me a bit about the proposed solution is that this seems to let the function do more than one specific task and we just rely on that particular task being a no-op for all use-cases other than when we actually want that to happen...

From a practical point of view, that'd work just fine, so this is a purely principal critique. However, given that this script is not expected to become part of a huge codebase, I think it'd be fine to use here.

So to sum it up: Personally, I wouldn't include this into that function, but I am not bothered too much by it either. So your call here.
However, if you decide to use the function as proposed, I think it warrants a comment stating that we expect the prefix-removal to be a no-op for most cases in which this function is called. Just to make everyone aware of this assumption.

d3adb5 committed:
refactor: use name normalizing function + sed for env vars

Add a function to "normalize" the name of config options that will be
used, while using sed instead of grep to both retrieve and manipulate
the names of environment variables that start with MUMBLE_CONFIG_.

d3adb5 commented:
> The only thing that itches me a bit about the proposed solution is that this seems to let the function do more than one specific task and we just rely on that particular task being a no-op for all use-cases other than when we actually want that to happen...

I can understand by principle it does exhibit multiple behaviors. Like you said, it isn't part of some huge codebase, and I would've left it as-is, but I think I managed to find a solution that is both brief and "principally moral" in [cf6a9fb](https://github.com/mumble-voip/mumble-docker/pull/6/commits/cf6a9fb4bf85abee7d1ae2d59251f1f8a53b298a):

```bash
# Does one thing, doesn't care about prefixes.
normalize_name() {
	local uppercase="${1^^}"
	echo "${uppercase//_/}"
}

# ...

while read ...; do
# ...
done < <( printenv --null | sed -zn 's/^MUMBLE_CONFIG_//p )
# ^ Retrieves vars and already filters out the MUMBLE_CONFIG_ prefix
```

Krzmbrzl commented:
Thank you for your contribution :+1:
